core/src/main/java/com/linecorp/armeria/internal/DefaultAttributeMap.java

```diff
@@ -263,11 +338,99 @@ private void remove0() {
         }
     }
 
+    private final class CopyOnWriteAttribute<T> implements Attribute<T> {
+
+        private final Attribute<T> attrInRoot;
+
+        @Nullable
+        private Attribute<T> attrInThis;
+
+        CopyOnWriteAttribute(Attribute<T> attrInRoot) {
+            this.attrInRoot = attrInRoot;
+        }
+
+        @Override
+        public AttributeKey<T> key() {
+            if (attrInThis != null) {
+                return attrInThis.key();
+            }
+            return attrInRoot.key();
+        }
+
+        @Override
+        public T get() {
+            if (attrInThis != null) {
+                return attrInThis.get();
+            }
+            return attrInRoot.get();
+        }
+
+        @Override
+        public void set(T value) {
+            if (attrInThis != null) {
+                attrInThis.set(value);
+            } else {
+                final Attribute<T> attrInThis = attr(attrInRoot.key(), true);
+                this.attrInThis = attrInThis;
+                attrInThis.set(value);
+            }
+        }
+
+        @Override
+        public T getAndSet(T value) {
+            if (attrInThis != null) {
+                return attrInThis.getAndSet(value);
+            }
+            set(value);
+            return attrInRoot.get();
+        }
+
+        @Override
+        public T setIfAbsent(T value) {
+            if (attrInThis != null) {
+                return attrInThis.setIfAbsent(value);
+            }
+            final Attribute<T> attrInThis = attr(attrInRoot.key(), true);
+            this.attrInThis = attrInThis;
+            return attrInThis.setIfAbsent(value);
+        }
+
+        @Override
+        public T getAndRemove() {
+            if (attrInThis != null) {
+                return attrInThis.getAndRemove();
+            }
+            return attrInRoot.get();
```

Comment:
Need copy value here? I expect the following result:

```java
final DefaultAttributeMap root = new DefaultAttributeMap(null);
final AttributeKey<String> rootKey = AttributeKey.valueOf("root-key");
root.attr(rootKey).set("ROOT");
final DefaultAttributeMap child = new DefaultAttributeMap(root);
assert child.attrs() instanceof CopyOnWriteAttributeIterator;
final Attribute<String> attribute = (Attribute<String>) child.attrs().next();
assert attribute.getAndRemove() == "ROOT";
assert attribute.getAndRemove() == null;  // assertion fail, returns "ROOT"
```

Reply:
Just call `getAndSet(null)` then own implementation?

Reply:
Nice catch, thanks!
Had a chat with @trustin and we need to revamp our attribute API.
Because the `AttributeMap` from netty has deprecated APIs such as `getAndRemove()` which will cause a race condition, we decided not to extend it anymore.
So we will add the following methods to `RequestContext`:
```
@Nullable
<T> T attr(AttributeKey<T> key);
<T> void setAttr(AttributeKey<T> key, T value);
Entry<AttributeKey<?>, ?> attrs();
```
and `ownX` methods to `ClientRequestContext`:
```
<T> T ownAttr(AttributeKey<T> key);
<T> void setOwnAttr(AttributeKey<T> key, T value);
Entry<AttributeKey<?>, ?> ownAttrs();
```
The reason why I don't define a new `AttributeMap` is that we are using the netty's `AttributeMap` in `ConnectionPoolListener` and didn't want to make a user get confused by using the same interface.

What do you think? @anuraaga

Reply:
Makes sense to me (less methods is always better when reasonable :) )